dqvm/src/global-addresses.lisp

```diff
@@ -0,0 +1,181 @@
+          (slot-value global-addresses 'remainder-blocks) rem)))
+
+(defmethod print-object ((global-addresses global-addresses) stream)
+  (let ((*print-readably* nil)
+        (*print-pretty* nil))
+
+    (print-unreadable-object (global-addresses stream :type t :identity t)
+
+      (format stream "~{~A ~A~^ ~}"
+              (list (prin1-to-string :number-of-qubits) (number-of-qubits global-addresses)
+                    (prin1-to-string :number-of-processes) (number-of-processes global-addresses)
+                    (prin1-to-string :block-size) (block-size global-addresses)
+                    (prin1-to-string :permutation) (permutation global-addresses))))))
+
+(defmethod copy ((global-addresses global-addresses))
+  (let ((number-of-qubits (number-of-qubits global-addresses))
+        (number-of-processes (number-of-processes global-addresses))
+        (block-size (block-size global-addresses))
+        (permutation (permutation global-addresses)))
+
+    (make-instance 'global-addresses :number-of-qubits number-of-qubits
+                                     :number-of-processes number-of-processes
+                                     :block-size block-size
+                                     :permutation permutation)))
+
+(defmethod equal-p ((g1 global-addresses) (g2 global-addresses))
+  (let ((number-of-qubits1 (number-of-qubits))
+        (number-of-processes1 (number-of-processes g1))
+        (block-size1 (block-size g1))
+        (permutation1 (permutation g1))
+
+        (number-of-qubits2 (number-of-qubits g2))
+        (number-of-processes2 (number-of-processes g2))
+        (block-size2 (block-size g2))
+        (permutation2 (permutation g2)))
+
+    (and (= number-of-qubits1 number-of-qubits2)
+         (= number-of-processes1 number-of-processes2)
+         (= block-size1 block-size2)
+         (equalp permutation1 permutation2))))
+
+(defmethod number-of-addresses ((global-addresses global-addresses))
+  (expt 2 (number-of-qubits global-addresses)))
+
+(defun %number-of-blocks (global-addresses)
+  "Return the total number of block matrices to consider in the simulation."
+  (ceiling (number-of-addresses global-addresses)
+           (block-size global-addresses)))
+
+(defmethod %blocks-per-process ((global-addresses global-addresses))
+  "Return the number of regular blocks per process in GLOBAL-ADDRESSES and the number of remaining blocks.
+
+Note that this may not be the number of blocks in the processes. To obtain that, you must check the rank and the remaining blocks."
+  (floor (%number-of-blocks global-addresses)
+         (number-of-processes global-addresses)))
+
+(defmethod get-initial-address ((global-addresses global-addresses) address)
+  "Apply the inverse of the latest permutation in ADDRESSES to ADDRESS."
+  (apply-inverse-qubit-permutation (permutation global-addresses) address))
+
+(defmethod get-block-by-address ((global-addresses global-addresses) address)
```

Comment:
What is the rationale for letting ADDRESS be NIL?

Reply:
That was a remnant of a previous interface. Thanks for catching it.